File: c#-Assembly-Line/AssemblyLine/BoundedBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AssemblyLine
{
    public class BoundedBuffer
    {
        public static int napTime = 5;
        public static int buffer_size = 3;
        private LinkedList<Widget> widgetItems = new LinkedList<Widget> ();
        private int numberOfItemsInBuffer;

        public BoundedBuffer()
        {
            numberOfItemsInBuffer = 0;
        }

        public static void napping()
        {
            Random rand = new Random();
            int sleepTime = napTime * rand.Next(10);
            try
            {
                Thread.Sleep(sleepTime);
            } catch (Exception e) { }
        }

        public async void enter(Widget item, string worker, string widgetNumber)
        {
            /*
            while (widgetItems.Count() == 3)
            {
                try
                {
                    Console.WriteLine($"{worker} encountered a full Buffer.");
                    //await Widget();
                } catch (Exception e) { }
                
            } 
            */

            numberOfItemsInBuffer++;

            widgetItems.AddLast(item);

            if(numberOfItemsInBuffer == buffer_size)
            {
                Console.WriteLine($"{worker} entered Widget {widgetNumber}. Buffer FULL.");
            } else
            {
                Console.WriteLine($"{worker} entered Widget {widgetNumber}. Buffer size is {numberOfItemsInBuffer}.");
            }
        }

        public Widget remove(string worker)
        {
            Widget item;

            /*
            while(widgetItems.Count() == 0)
            {
                Console.WriteLine($"{worker} is waiting.");
            }
            */

            numberOfItemsInBuffer--;
            item = widgetItems.First();
            widgetItems.RemoveFirst();

            return item;
        }

        public void finalList(Widget item, string worker, int widgetNumber)
        {
            widgetItems.AddLast(item);
        }

        public void print()
        {
            int widgetNum = 1;
            foreach(Widget entry in widgetItems)
            {
                Console.WriteLine("--------------------");
                Console.WriteLine($"Widget # {widgetNum}");
                Console.WriteLine($"Widget ID: {entry.getModelNumber}");
                Console.WriteLine($"Number of Workers that Processed {entry.numberOfWorkers}");
                widgetNum++;
            }
        }
    }
}

File: Java-Production-Line/src/BoundedBuffer.java
import java.util.LinkedList;

public class BoundedBuffer {

    // VARIABLES
    public static final int NAP_TIME = 5;
    private static final int BUFFER_SIZE = 3;
    public String bufferName;
    private LinkedList<Widget> widgetItems = new LinkedList<>(); // CREATING THE LINKED LIST
    private int numberOfItemsInBuffer; // NUMBER OF ITEMS IN THE BUFFER

    // CONSTRUCTOR
    public BoundedBuffer(String name) {
        numberOfItemsInBuffer = 0;
        bufferName = name;
    }

    // METHOD TO SLEEP THE THREAD FOR A PERIOD
    public static void napping() {
        int sleepTime = (int) (NAP_TIME * Math.random() );
        try {
            Thread.sleep(sleepTime*800); }
        catch(InterruptedException e) { }
    }

    // ANIMATION METHOD FOR ADDING NEW CUBES TO THE SCREEN
    public void enterScreen(AnimationCube cube) {
        switch (numberOfItemsInBuffer) {
            case 0:
                cube.setPosition(404);
                break;
            case 1:
                cube.setPosition(299);
                break;
            case 2:
                cube.setPosition(194);
                break;
        }
    }

    // METHOD TO ENTER ITEMS ONTO THE BUFFER
    public synchronized void enter(Widget item, String worker, String widgetNumber, AnimationCube cube, int unitNumber) {
        // CHECKS TO SEE IF THE BUFFER IS FULL
        while (widgetItems.size() == 3) {
            try {
                wait();
            } catch (InterruptedException e) { }
        }

        // INCREMENTS BUFFER COUNT
        numberOfItemsInBuffer++;

        // ADDS WIDGET TO BUFFER LIST
        widgetItems.add(item);

        // PROGRESS OUTPUT
        if(numberOfItemsInBuffer == BUFFER_SIZE) {
            System.out.println(worker + " entered Widget " + widgetNumber + " onto Belt " + bufferName + ". Buffer FULL.");
        } else {
            System.out.println(worker + " entered Widget " + widgetNumber + " onto Belt " + bufferName + ". Buffer size is " + numberOfItemsInBuffer);
        }

        // NOTIFIES NEXT THREAD
        notify();
    }

    // ANIMATION METHOD FOR MOVING CUBES FROM THE FIRST BELT TO THE SECOND BELT
    public void secondBelt(AnimationCube cube) {
        switch (numberOfItemsInBuffer) {
            case 0:
                cube.setPosition(825);
                break;
            case 1:
                cube.setPosition(720);
                break;
            case 2:
                cube.setPosition(615);
                break;
        }
    }

    // ANIMATION METHOD FOR MOVING CUBES FROM THE SECOND BELT TO THE THIRD BELT
    public void thirdBelt(AnimationCube cube) {
        switch (numberOfItemsInBuffer) {
            case 0:
                cube.setPosition(1246);
                break;
            case 1:
                cube.setPosition(1141);
                break;
            case 2:
                cube.setPosition(1036);
                break;
        }
    }

    // METHOD FOR REMOVING ITEMS FROM THE BUFFER
    public synchronized Widget remove(String worker) {
        Widget item;

        // CHECKS TO SEE IF THE INPUT LIST IS EMPTY
        while(widgetItems.size() == 0) {
            System.out.println(worker + " is waiting.");
            try {
                wait();
            } catch (InterruptedException e) { }
        }

        numberOfItemsInBuffer--;
        item = widgetItems.removeFirst();
        notify();
        return item;
    }

    // METHOD FOR ADDING THE COMPLETED ITEMS TO THE FINAL OUTPUT LIST
    public synchronized  void finalList(Widget item, String worker, int widgetNumber) {
        widgetItems.add(item);
        System.out.println(worker + " completed work on Widget " + widgetNumber + ". Widget exiting production line.");
    }


    // METHOD TO PRINT THE FINAL RESULT
    public void print() {
        for(int i = 0; i < widgetItems.size(); i++) {
            int widgetNum = i+1;
            System.out.println("--------------------------------------------");
            System.out.println("Widget # " + widgetNum);
            System.out.println("Widget ID: " + widgetItems.get(i).getModelNumber());
            System.out.println("Number of Workers that Processed: " + widgetItems.get(i).numberOfWorkers);
        }
    }
}
